contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1049 @@
+    _ERC1820_REGISTRY.setInterfaceImplementer(
+      address(this),
+      ERC777_TOKENS_RECIPIENT_HASH,
+      address(this)
+    );
+  }
+
+  // View functions and getters =========================================
+  function supportsInterface(bytes4 interfaceId)
+    public
+    view
+    override(ERC1155Upgradeable, AccessControlEnumerableUpgradeable)
+    returns (bool)
+  {
+    return super.supportsInterface(interfaceId);
+  }
+
+  /**
+   * Returns the schedule duration in seconds that has been set for a given methodology and
+   * methodology version.
+   */
+  function getRestrictionDurationForMethodologyAndVersion(
+    uint256 methodology,
+    uint256 methodologyVersion
+  ) public view returns (RestrictionDuration memory) {
+    return
+      _methodologyAndVersionToScheduleDuration[methodology][methodologyVersion];
+  }
+
+  /** Returns an array of all existing schedule ids, regardless of the status of the schedule. */
+  function getAllScheduleIds() external view returns (uint256[] memory) {
+    uint256[] memory allScheduleIdsArray = new uint256[](
+      _allScheduleIds.length()
+    );
+    for (uint256 i = 0; i < allScheduleIdsArray.length; i++) {
+      allScheduleIdsArray[i] = _allScheduleIds.at(i);
+    }
+    return allScheduleIdsArray;
+  }
+
+  /** Returns an array of all schedule ids of which an address currently owns any tokens. */
+  function getScheduleIdsForAccount(address account)
+    external
+    view
+    returns (uint256[] memory)
+  {
+    EnumerableSetUpgradeable.UintSet
+      storage scheduleIdSet = _addressToScheduleIdSet[account];
+    return _enumerableSetToArray(scheduleIdSet);
+  }
+
+  /** Returns an account-specific view of the details of a specific schedule. */
+  function getScheduleDetailForAccount(address account, uint256 scheduleId)
+    public
+    view
+    returns (ScheduleDetailForAddress memory)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+
+    return
+      ScheduleDetailForAddress(
+        account,
+        scheduleId,
+        schedule.startTime,
+        schedule.endTime,
+        balanceOf(account, scheduleId),
+        claimableBalanceForScheduleForAccount(scheduleId, account),
+        schedule.claimedAmountsByAddress[account],
+        schedule.quantitiesRevokedByAddress[account],
+        schedule.exists
+      );
+  }
+
+  /** Returns an account-specific view of the details of all schedules that an account has ownership of. */
+  function batchGetScheduleDetailsForAccount(address account)
+    external
+    view
+    returns (ScheduleDetailForAddress[] memory)
+  {
+    EnumerableSetUpgradeable.UintSet
+      storage scheduleIds = _addressToScheduleIdSet[account];
+    ScheduleDetailForAddress[]
+      memory scheduleDetails = new ScheduleDetailForAddress[](
+        scheduleIds.length()
+      );
+    for (uint256 i = 0; i < scheduleIds.length(); i++) {
+      scheduleDetails[i] = getScheduleDetailForAccount(
```

Comment:
should this have a check to see if `scheduleIds.at(i)` exists before calling it? Otherwise this function is all or nothing if one account doesn't have a schedule

Reply:
particularly because I think [this line](https://github.com/nori-dot-eco/contracts/pull/249/files#diff-de34eb981fc0fa6e6b114737c58140e5e1b03e6f25139d619de635e4e0a9d7e0R482) could cause an underflow in loading the schedule and revert

Reply:
it should never be the case that a schedule doesn't exist if the schedule id is in the schedule id for an address (`_addressToScheduleIdSet`), so they should always exist but since this is a view function maybe the extra check isn't a bad idea so that this whole call doesn't revert if the schedule somehow doesn't actually exist

Reply:
hmm that line doesn't seem to be linking to a specific line

Reply:
You might be right-- thought there was a possibility some zero value could be passed into `claimableBalanceForScheduleForAccount` but not seeing where I thought I saw that now